nnfabrik/templates/transfer/transferred_trained_model.py

```diff
@@ -0,0 +1,310 @@
+        ret = super().get_full_config(
+            key=key,
+            include_state_dict=include_state_dict,
+            include_trainer=include_trainer,
+        )
+        prev_key = self.CollapsedHistory().get_prev_key(key)
+        # retrieve corresponding model state (and overwrite possibly retrieved state)
+        if include_state_dict and (self.ModelStorage & prev_key):
+            with tempfile.TemporaryDirectory() as temp_dir:
+                state_dict_path = (self.ModelStorage & prev_key).fetch1("model_state", download_path=temp_dir)
+                ret["state_dict"] = torch.load(state_dict_path)
+                ret["model_config"]["transfer"] = True
+        # retrieve data if present (if previous step did data transfer)
+        if self.DataStorage & prev_key:
+            with tempfile.TemporaryDirectory() as temp_dir:
+                data_path = (self.DataStorage & prev_key).fetch1("transfer_data", download_path=temp_dir)
+                ret["dataset_config"]["transfer_data"] = np.load(data_path)
+        return ret
+
+    def make(self, key):
+        """
+        Given key specifying configuration for dataloaders, model and trainer,
+        trains the model and saves the trained model.
+        """
+
+        # lookup the fabrikant corresponding to the current DJ user
+        fabrikant_name = self.user_table.get_current_user()
+        seed = (self.seed_table & key).fetch1("seed")
+
+        # load everything
+        dataloaders, model, trainer = self.load_model(key, include_trainer=True, include_state_dict=True, seed=seed)
+
+        # define callback with pinging
+        def call_back(**kwargs):
+            self.connection.ping()
+            self.call_back(**kwargs)
+
+        trainer_output = trainer(model=model, dataloaders=dataloaders, seed=seed, uid=key, cb=call_back)
+        score, output, model_state = trainer_output[:3]
+        transfer_data = {} if len(trainer_output) < 4 else trainer_output[3]
+
+        with tempfile.TemporaryDirectory() as temp_dir:
+            filename = make_hash(key)
+            key["score"] = score
+            key["output"] = output
+            key["fabrikant_name"] = fabrikant_name
+            comments = []
+            comments.append((self.trainer_table & key).fetch1("trainer_comment"))
+            comments.append((self.model_table & key).fetch1("model_comment"))
+            comments.append((self.dataset_table & key).fetch1("dataset_comment"))
+            key["comment"] = self.comment_delimitter.join(comments)
+
+            self.insert1(key)
+            self.CollapsedHistory.add_entry(key)
+
+            prev_key = self.CollapsedHistory().get_prev_key(key)
+            if key["data_transfer"]:
+                if transfer_data:
+                    data_path = os.path.join(temp_dir, filename + "_transfer_data.npz")
+                    np.savez(data_path, **transfer_data)
+                elif self.DataStorage & prev_key:
```

Comment:
So if the current step doesn't add a new `transfer_data`, you reuse the `transfer_data` from the previous transfer step?

Reply:
That was the idea. I updated this to only save at the step where the data is produced and instead go through the history as far as necessary when loading the data.